From e60ae715ac0515a9729f0922cafe711a53afddc7 Mon Sep 17 00:00:00 2001
From: Stephen Nusko <nuskos@chromium.org>
Date: Thu, 29 May 2025 00:17:26 -0700
Subject: [PATCH] Use std::array instead of a c style array in PurgePartitionAlloc

Currently if an attacker has somehow corrupted the freelist we might
return a index beyond kMaxSlotCount which could be a OOB write. However
std::array is hardened in chromium so it does bounds checking preventing
this exploit path.

Separate work is thinking about hardening the GetNext function of
freelist, but this is a quick and easy mitigation.

Bug: 331454364
Change-Id: I4154a4fd5b373c8ecda21d93257eda4beaa419cb
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6596688
Reviewed-by: Keishi Hattori <keishi@chromium.org>
Auto-Submit: Stephen Nusko <nuskos@chromium.org>
Commit-Queue: Keishi Hattori <keishi@chromium.org>
Commit-Queue: Stephen Nusko <nuskos@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1466906}
---

--- a/base/allocator/partition_allocator/src/partition_alloc/partition_root.cc
+++ b/base/allocator/partition_allocator/src/partition_alloc/partition_root.cc
@@ -4,6 +4,8 @@
 
 #include "partition_alloc/partition_root.h"
 
+#include <algorithm>
+#include <array>
 #include <cstdint>
 
 #include "partition_alloc/bucket_lookup.h"
@@ -490,13 +492,13 @@
   PA_DCHECK(slot_span->num_unprovisioned_slots < bucket_num_slots);
   size_t num_provisioned_slots =
       bucket_num_slots - slot_span->num_unprovisioned_slots;
-  char slot_usage[kMaxSlotCount];
+  std::array<char, kMaxSlotCount> slot_usage{};
 #if !PA_BUILDFLAG(IS_WIN)
   // The last freelist entry should not be discarded when using OS_WIN.
   // DiscardVirtualMemory makes the contents of discarded memory undefined.
   size_t last_slot = static_cast<size_t>(-1);
 #endif
-  memset(slot_usage, 1, num_provisioned_slots);
+  std::fill_n(slot_usage.begin(), num_provisioned_slots, 1);
   uintptr_t slot_span_start = internal::SlotSpanMetadata<
       internal::MetadataKind::kReadOnly>::ToSlotSpanStart(slot_span);
   // First, walk the freelist for this slot span and make a bitmap of which
